provider/digital_ocean.go

```diff
@@ -0,0 +1,279 @@
+	ResourceRecordSet godo.DomainRecord
+}
+
+const (
+	// DigitalOceanCreate is a ChangeAction enum value
+	DigitalOceanCreate = "CREATE"
+	// DigitalOceanDelete is a ChangeAction enum value
+	DigitalOceanDelete = "DELETE"
+	// DigitalOceanUpdate is a ChangeAction enum value
+	DigitalOceanUpdate = "UPDATE"
+)
+
+// Token returns oauth2 token struct in order to use Digital Ocean's API.
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+// NewDigitalOceanProvider initializes a new DigitalOcean DNS based Provider.
+func NewDigitalOceanProvider(domainFilter string, dryRun bool) (Provider, error) {
+	token := os.Getenv("DO_TOKEN")
+	tokenSource := &TokenSource{
+		AccessToken: token,
+	}
+	if len(token) == 0 {
+		return nil, fmt.Errorf("No token found")
+	}
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	client := godo.NewClient(oauthClient)
+
+	provider := &DigitalOceanProvider{
+		Client:       client.Domains,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}
+	return provider, nil
+}
+
+// Zones returns the list of hosted zones.
+func (p *DigitalOceanProvider) Zones() ([]godo.Domain, error) {
+	result := []godo.Domain{}
+
+	zones, _, err := p.Client.List(context.TODO(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		if strings.HasSuffix(zone.Name, p.domainFilter) {
+			result = append(result, zone)
+		}
+	}
+
+	return result, nil
+}
+
+// Records returns the list of records in a given zone.
+func (p *DigitalOceanProvider) Records() ([]*endpoint.Endpoint, error) {
+	zones, err := p.Zones()
+	if err != nil {
+		return nil, err
+	}
+	endpoints := []*endpoint.Endpoint{}
+	for _, zone := range zones {
+		records, _, err := p.Client.Records(context.TODO(), zone.Name, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range records {
+			endpoints = append(endpoints, endpoint.NewEndpoint(r.Name, r.Data, r.Type))
+		}
+	}
+
+	return endpoints, nil
+}
+
+// submitChanges takes a zone and a collection of Changes and sends them as a single transaction.
+func (p *DigitalOceanProvider) submitChanges(changes []*DigitalOceanChange) error {
+	// return early if there is nothing to change
+	if len(changes) == 0 {
+		return nil
+	}
+
+	zones, err := p.Zones()
+	if err != nil {
+		return err
+	}
+
+	// separate into per-zone change sets to be passed to the API.
+	changesByZone := digitalOceanChangesByZone(zones, changes)
+
+	for zoneName, changes := range changesByZone {
+		for _, change := range changes {
+			logFields := log.Fields{
+				"record": change.ResourceRecordSet.Name,
+				"type":   change.ResourceRecordSet.Type,
+				"action": change.Action,
+				"zone":   zoneName,
+			}
+
+			log.WithFields(logFields).Info("Changing record.")
+
+			if p.DryRun {
+				continue
+			}
+			switch change.Action {
+			case DigitalOceanCreate:
+				_, _, err := p.Client.CreateRecord(context.TODO(), zoneName,
```

Comment:
that's exactly the reason why we didn't go wtih ConfigMap implementation for ownership yet :)